Error:  mktemp: failed to create directory via template

Hello,

I'm rebuilding some packages of the  cloudposse/packages container and I got this error:  mktemp: failed to create directory via template ‘../../tmp/tmp.XXXXXXXXXX’: No such file or directory

I could observe that the problem is related to this file included at makefile: "/packages/tasks/Makefile.apk"

...
export APK_TMP_DIR := $(realpath $(shell mktemp -p ../../tmp -d))
...

/packages # echo "1.14.0" > /packages/vendor/kubectl/VERSION
/packages # make -C install kubectl
make: Entering directory '/packages/install'
make[1]: Entering directory '/packages/vendor/kubectl'
mktemp: failed to create directory via template ‘../../tmp/tmp.XXXXXXXXXX’: No such file or directory
curl --retry 3 --retry-delay 5 --fail -sSL -o /packages/bin/kubectl https://storage.googleapis.com/kubernetes-release/release/v1.14.0
/bin/linux/amd64/kubectl && chmod +x /packages/bin/kubectl
make[1]: Leaving directory '/packages/vendor/kubectl'
make: Leaving directory '/packages/install'

Hello @osterman,  I'm using cloudposse/package:latest as the base image for my docker build and I still getting the same error:

Dockerfile:

#
# Cloud Posse Package Distribution
#
FROM cloudposse/packages:latest as packages

WORKDIR /packages

#
# Install the select packages from the cloudposse package manager image
#
# Repo: <https://github.com/cloudposse/packages>
#

ENV TERRAFORM_VERSION   ${TERRAFORM_VERSION:-0.11.13}
ENV PACKER_VERSION      ${PACKER_VERSION:-1.3.5}
ENV KUBE_VERSION        ${KUBE_VERSION:-1.14.0}
ENV HELM_VERSION        ${HELM_VERSION:-2.13.1}
ENV HELMFILE_VERSION    ${HELM_DIFF_VERSION:-0.51.1}

RUN echo ${TERRAFORM_VERSION} > /packages/vendor/terraform/VERSION \
&& echo ${PACKER_VERSION}    > /packages/vendor/packer/VERSION \
&& echo ${KUBE_VERSION}      > /packages/vendor/kubectl/VERSION \
&& echo ${HELM_VERSION}      > /packages/vendor/helm/VERSION \
&& echo ${HELMFILE_VERSION}  > /packages/vendor/helmfile/VERSION

ARG PACKAGES="terraform packer kubectl helm helmfile"
ENV PACKAGES=${PACKAGES}
RUN make -C install $PACKAGES \
&& make dist


BUILDLOG:

docker build -t mypackages .
Sending build context to Docker daemon  3.072kB
Step 1/11 : FROM cloudposse/packages:latest as packages
latest: Pulling from cloudposse/packages
8e402f1a9c57: Already exists 
e8e2d1e0cafa: Pull complete 
74011f0ed35a: Pull complete 
aecce3c93e5c: Pull complete 
f9309f958424: Pull complete 
34ffc0c23ecd: Pull complete 
Digest: sha256:72d85b0af1d135afb971d3806b31ee2a674d27748b08933f7658bee533057019
Status: Downloaded newer image for cloudposse/packages:latest
 ---> e2f833477714
Step 2/11 : WORKDIR /packages
 ---> Running in 0249724beac0
Removing intermediate container 0249724beac0
 ---> 906c2bb5d773
Step 3/11 : ENV TERRAFORM_VERSION   ${TERRAFORM_VERSION:-0.11.13}
 ---> Running in 7282b3e770c4
Removing intermediate container 7282b3e770c4
 ---> c623990175e5
Step 4/11 : ENV PACKER_VERSION      ${PACKER_VERSION:-1.3.5}
 ---> Running in d3f438c8388f
Removing intermediate container d3f438c8388f
 ---> 1cfa096cba5b
Step 5/11 : ENV KUBE_VERSION        ${KUBE_VERSION:-1.14.0}
 ---> Running in a537a8a1deec
Removing intermediate container a537a8a1deec
 ---> a416aa59492e
Step 6/11 : ENV HELM_VERSION        ${HELM_VERSION:-2.13.1}
 ---> Running in 1a0cffdda219
Removing intermediate container 1a0cffdda219
 ---> 8c231a04d0f1
Step 7/11 : ENV HELMFILE_VERSION    ${HELM_DIFF_VERSION:-0.51.1}
 ---> Running in 057ed68855be
Removing intermediate container 057ed68855be
 ---> b1be1874c91d
Step 8/11 : RUN echo ${TERRAFORM_VERSION} > /packages/vendor/terraform/VERSION && echo ${PACKER_VERSION}    > /packages/vendor/packer/VERSION && echo ${KUBE_VERSION}      > /packages/vendor/kubectl/VERSION && echo ${HELM_VERSION}      > /packages/vendor/helm/VERSION && echo ${HELMFILE_VERSION}  > /packages/vendor/helmfile/VERSION
 ---> Running in cc4c2ae5062f
Removing intermediate container cc4c2ae5062f
 ---> c08b54c8fb7b
Step 9/11 : ARG PACKAGES="terraform packer kubectl helm helmfile"
 ---> Running in 76cda1e9282f
Removing intermediate container 76cda1e9282f
 ---> 33628e2f4729
Step 10/11 : ENV PACKAGES=${PACKAGES}
 ---> Running in 0c2d1eaa1736
Removing intermediate container 0c2d1eaa1736
 ---> 1569d7270a4f
Step 11/11 : RUN make -C install $PACKAGES && make dist
 ---> Running in faf4c4a25dbb
make: Entering directory '/packages/install'
make[1]: Entering directory '/packages/vendor/terraform'
mktemp: failed to create directory via template ‘../../tmp/build.XXXXXX’: No such file or directory
mkdir -p /tmp/terraform
curl --retry 3 --retry-delay 5 --fail -sSL -o /tmp/terraform/terraform.zip https://releases.hashicorp.com/terraform/0.11.13/terraform_0.11.13_linux_amd64.zip
unzip -d /tmp/terraform /tmp/terraform/terraform.zip
Archive:  /tmp/terraform/terraform.zip
  inflating: terraform
mv /tmp/terraform/terraform /packages/bin/terraform
chmod +x /packages/bin/terraform
make[1]: Leaving directory '/packages/vendor/terraform'
make[1]: Entering directory '/packages/vendor/packer'
mktemp: failed to create directory via template ‘../../tmp/build.XXXXXX’: No such file or directory
curl --retry 3 --retry-delay 5 --fail -sSL -o - https://releases.hashicorp.com/packer/1.3.5/packer_1.3.5_linux_amd64.zip > /packages/bin/packer.zip
unzip -p /packages/bin/packer.zip packer > /packages/bin/packer
rm /packages/bin/packer.zip
chmod +x /packages/bin/packer
make[1]: Leaving directory '/packages/vendor/packer'
make[1]: Entering directory '/packages/vendor/kubectl'
mktemp: failed to create directory via template ‘../../tmp/build.XXXXXX’: No such file or directory
curl --retry 3 --retry-delay 5 --fail -sSL -o /packages/bin/kubectl https://storage.googleapis.com/kubernetes-release/release/v1.14.0/bin/linux/amd64/kubectl && chmod +x /packages/bin/kubectl
make[1]: Leaving directory '/packages/vendor/kubectl'
make[1]: Entering directory '/packages/vendor/helm'
mktemp: failed to create directory via template ‘../../tmp/build.XXXXXX’: No such file or directory
mkdir -p /tmp/helm
curl --retry 3 --retry-delay 5 --fail -sSL -o - http://storage.googleapis.com/kubernetes-helm/helm-v2.13.1-linux-amd64.tar.gz | tar -C /tmp/helm -zx linux-amd64/helm
curl: (7) Failed to connect to storage.googleapis.com port 80: Connection refused
gzip: invalid magic
tar: Child returned status 1
tar: Error is not recoverable: exiting now
make[1]: *** [Makefile:11: install] Error 2
make[1]: Leaving directory '/packages/vendor/helm'
make: *** [Makefile:9: helm] Error 2
make: Leaving directory '/packages/install'
The command '/bin/sh -c make -C install $PACKAGES && make dist' returned a non-zero code: 2

--- a/.dockerignore
+++ b/.dockerignore
@@ -8,4 +8,5 @@
 *.iml
 .build-harness
 build-harness
-tmp
+# We are relying on tmp and must include it
+tmp/**

--- a/Dockerfile
+++ b/Dockerfile
@@ -12,5 +12,5 @@ COPY --from=cfssl /go/bin/ ${INSTALL_PATH}/
 
 COPY . /packages
 RUN mkdir -p /packages/tmp
-RUN make -C /packages/install/ terraform
+RUN make -C /packages/install/ all
 WORKDIR /packages

--- a/Dockerfile
+++ b/Dockerfile
@@ -10,6 +10,7 @@ RUN apk add --update --no-cache bash make curl coreutils libc6-compat tar xz
 
 COPY --from=cfssl /go/bin/ ${INSTALL_PATH}/
 
-ADD . /packages
-RUN make -C /packages/install/ all
+COPY . /packages
+RUN mkdir -p /packages/tmp
+RUN make -C /packages/install/ terraform
 WORKDIR /packages

--- a/tasks/Makefile.apk
+++ b/tasks/Makefile.apk
@@ -26,6 +26,7 @@ apk/build:
 apk/clean:
 	@find $(APK_PACKAGES_PATH) -type f -name APKINDEX.tar.gz -delete
 	rm -rf $(APK_TMP_DIR)
+	mkdir -p $(APK_TMP_DIR)
 
 apk: apk/prepare apk/checksum apk/build apk/clean
 	@exit 0

